src/Compilers/CSharp/Portable/Binder/PatternSwitchBinder.cs

```diff
@@ -25,38 +26,132 @@ internal PatternSwitchBinder(Binder next, SwitchStatementSyntax switchSyntax) :
+            bool hasErrors = CheckSwitchErrors(node, boundSwitchGoverningExpression, ref switchSections, decisionDag, diagnostics);
+
+            if (boundSwitchGoverningExpression.ConstantValue != null)
+            {
+                // When the input is constant, we use that to reshape the decision dag that is returned
+                // so that flow analysis will see that some of the cases may be unreachable.
+                decisionDag = decisionDag.SimplifyDecisionDagForConstantInput(boundSwitchGoverningExpression, Conversions, diagnostics);
+            }
+
             return new BoundPatternSwitchStatement(
-                node, boundSwitchGoverningExpression, someCaseMatches,
-                locals, functions, switchSections, defaultLabel, this.BreakLabel, isComplete);
+                syntax: node,
+                expression: boundSwitchGoverningExpression,
+                innerLocals: locals,
+                innerLocalFunctions: functions,
+                switchSections: switchSections,
+                defaultLabel: defaultLabel,
+                breakLabel: this.BreakLabel,
+                decisionDag: decisionDag,
+                hasErrors: hasErrors);
+        }
+
+        private bool CheckSwitchErrors(
+            SwitchStatementSyntax node,
+            BoundExpression boundSwitchGoverningExpression,
+            ref ImmutableArray<BoundPatternSwitchSection> switchSections,
+            BoundDecisionDag decisionDag,
+            DiagnosticBag diagnostics)
+        {
+            HashSet<LabelSymbol> reachableLabels = decisionDag.ReachableLabels;
+            bool areAnySubsumed(ImmutableArray<BoundPatternSwitchSection> sections)
+            {
+                foreach (BoundPatternSwitchSection section in sections)
+                {
+                    foreach (BoundPatternSwitchLabel label in section.SwitchLabels)
+                    {
+                        if (!label.HasErrors && !reachableLabels.Contains(label.Label))
+                        {
+                            // we found a label that is subsumed
+                            return true;
+                        }
+                    }
+                }
+
+                return false;
+            }
+
+            if (!areAnySubsumed(switchSections))
+            {
+                return false;
+            }
+
+            // We mark any subsumed sections as erroneous for the benefit of flow analysis
+            var sectionBuilder = ArrayBuilder<BoundPatternSwitchSection>.GetInstance();
+            foreach (var oldSection in switchSections)
+            {
+                var labelBuilder = ArrayBuilder<BoundPatternSwitchLabel>.GetInstance();
+                foreach (var label in oldSection.SwitchLabels)
+                {
+                    var newLabel = label;
+                    if (!label.HasErrors && !reachableLabels.Contains(label.Label) && label.Syntax.Kind() != SyntaxKind.DefaultSwitchLabel)
```

Comment:
>!label.HasErrors && !reachableLabels.Contains(label.Label) [](start = 24, length = 58)

Consider extracting an `isSubsumed()` local function, for use here and in `areAnySubsumed()` above.